File: Data Structures and Algorithms/Python/LinkedList.py
class Node:
  value = None
  nextNode = None

  def __init__(self, value):
    self.value = value


class LinkedList:
  first = None

  def countNodes(self):
    if self.first == None:
      return 0

    count = 0
    save = self.first
    while (save != None):
      count += 1
      save = save.nextNode

    return count

  def getNodeAt(self, element):
    totalCount = self.countNodes()

    if element > totalCount or element < 1:
      print("Element doesn't exist")
      return
    
    if self.first == None:
      return 0
      
    save = self.first
    for i in range(1, element):
      save = save.nextNode

    return save
  
  def __sortLinkedList(self):
    if self.first == None:
      print("Linkedlist Underflow")
      return
    
    for i in range(1, self.countNodes()):
      for j in range(1, self.countNodes() - i):
        if self.getNodeAt(j).value > self.getNodeAt(j + 1).value:
          self.getNodeAt(j).value, self.getNodeAt(j + 1).value = self.getNodeAt(j + 1).value, self.getNodeAt(j).value
    
 
  def display(self):
    save = self.first
    while (save != None):
      print("Value:", save.value, "Next Node's Address:", save.nextNode)
      save = save.nextNode

  # mode 0 for insert at beginning
  # mode 1 for insert at end
  # mode 2 for insert in between (Only for insert)
  
  def insert(self, value, mode):
    node = Node(value)
    match mode:
      case 0:
        if self.first == None:
          self.first = node
        else:
          node.nextNode = self.first
          self.first = node
  
      case 1:
        if self.first == None:
          self.first = node
        else:
          save = self.first
          while(save.nextNode != None):
            save = save.nextNode
          save.nextNode = node
          
      case 2:
        if self.first == None:
          self.first = node
        else:
          save = self.first
          self.__sortLinkedList()
          for i in range(1, self.countNodes()):
            if(save.nextNode.value < value):
              save = save.nextNode
              
          save.nextNode, node.nextNode = node, save.nextNode

  def contains(self, value):
    for i in range(1, self.countNodes() + 1):
      if(value == self.getNodeAt(i).value):
        hasValue = True
        break
      else:
        hasValue = False

    return hasValue

  def remove(self, mode):
    if self.first == None:
      print("Linkedlist Underflow")
      return

    match mode:
      case 0:
        self.first = self.first.nextNode
        return

      case 1:
        save = self.first
        for i in range(1, self.countNodes() - 1):
          save = save.nextNode

        save.nextNode = None
        return

  print("Invalid Mode")

  def orderedRemove(self, value):
    if self.first == None:
      print("Linkedlist Underflow")
      return
    elif self.contains(value) == False:
      print("Element doesn't exist")
      return

    self.__sortLinkedList()
    save = self.first

    if save.value == value:
      self.remove(0)
      return
    
    for i in range(1, self.countNodes()):
      if save.nextNode.value == value:
        break
      else:
        save = save.nextNode

    tmp = save.nextNode.nextNode
    save.nextNode = tmp
    

File: Data Structures and Algorithms/C++/LinkedList.cpp
/*
    Ayush Kumar (202103103510253)
    4th Sem Div: A
    LinkedList.cpp
*/

#include "LinkedList.hpp"
#include <exception>
#include <iostream>

using namespace std;

void LinkedList::display() {
  Node *first = this->first;
  while (first != nullptr) {
    cout << "Value: " << first->value
         << " Next Node's Address: " << first->nextNode << endl;
    first = first->nextNode;
  }
}

int LinkedList::countNodes() {
  int count = 0;

  Node *save = this->first;

  while (save != nullptr) {
    count++;
    save = save->nextNode;
  }

  return count;
}

void LinkedList::insert(int value, int mode) {
  Node *node = new Node{value, nullptr};

  switch (mode) {
  case 0:
    if (this->first == nullptr) {
      this->first = node;
    } else {
      node->nextNode = this->first;
      this->first = node;
    }

    break;

  case 1:
    if (this->first == nullptr) {
      this->first = node;
    } else {
      Node *save = this->first;

      while (save->nextNode != nullptr) {
        save = save->nextNode;
      }
      save->nextNode = node;
    }
    break;

  case 2:
    if (this->first == nullptr) {
      this->first = node;
    } else {
      Node *save = this->first;
      LinkedList::sortLinkedList();

      for (int _ = 1; _ < LinkedList::countNodes(); _++) {
        if (save->nextNode->value < value) {
          save = save->nextNode;
        }
      }

      Node *tmp = save->nextNode;
      save->nextNode = node;
      node->nextNode = tmp;
    }
    break;
  default:
    cout << "Invalid Mode" << endl;
  }
}

Node &LinkedList::getNodeAt(int element) {
  int totalElements = LinkedList::countNodes();
  if (element > totalElements || element < 1) {
    throw invalid_argument("Element not found");
  }

  Node *save = this->first;

  for (int _ = 1; _ < element; _++) {
    save = save->nextNode;
  }

  return *save;
}

void LinkedList::remove(int mode) {
  if (this->first == nullptr) {
    cout << "LinkedList underflow" << endl;
    return;
  }

  switch (mode) {

  case 0:
    this->first = first->nextNode;
    break;

  case 1: {
    Node *save = this->first;
    for (int _ = 1; _ < LinkedList::countNodes() - 1; _++) {
      save = save->nextNode;
    }

    delete save->nextNode;
    save->nextNode = nullptr;
    break;
  }

  default:
    cout << "Invalid Mode" << endl;
  }
}

void LinkedList::sortLinkedList() {
  if (this->first == nullptr) {
    cout << "LinkedList underflow" << endl;
    return;
  }

  for (int i = 1; i <= LinkedList::countNodes(); i++) {
    for (int j = 1; j <= LinkedList::countNodes() - 1 - i; j++) {
      if (LinkedList::getNodeAt(j).value > LinkedList::getNodeAt(j + 1).value) {

        int tmp = LinkedList::getNodeAt(j).value;
        LinkedList::getNodeAt(j).value = LinkedList::getNodeAt(j + 1).value;
        LinkedList::getNodeAt(j + 1).value = tmp;
      }
    }
  }
}

void LinkedList::orderedRemove(int value) {
  if (this->first == nullptr) {
    cout << "LinkedList underflow" << endl;
    return;
  } else if (!LinkedList::contains(value)) {
    cout << "Element Doesn't exist" << endl;
    return;
  }

  LinkedList::sortLinkedList();

  Node *save = this->first;

  if (save->value == value) {
    LinkedList::remove(0);
    return;
  }

  for (int _ = 1; _ <= LinkedList::countNodes(); _++) {
    if (save->nextNode->value < value) {
      save = save->nextNode;
    }
  }

  Node *tmp = save->nextNode->nextNode;
  delete save->nextNode;
  save->nextNode = tmp;
}

bool LinkedList::contains(int value) {
  bool hasValue = false;

  for (int i = 1; i <= LinkedList::countNodes(); i++) {
    if (LinkedList::getNodeAt(i).value == value) {
      hasValue = true;
      break;
    }
  }
  return hasValue;
}